src/DataProtection/DataProtection/src/KeyManagement/KeyRingProvider.cs

```diff
@@ -164,7 +164,24 @@ private CacheableKeyRing CreateCacheableKeyRingCoreStep2(DateTimeOffset now, Can
 
         _logger.UsingKeyAsDefaultKey(defaultKey.KeyId);
 
-        var nextAutoRefreshTime = now + GetRefreshPeriodWithJitter(KeyManagementOptions.KeyRingRefreshPeriod);
+        // If we're in a scenario where we think other app instances may be generating keys that will be
+        // activated before they are propagated, we'll use a shorter refresh period to ensure that we pick
+        // up those keys so that we can select from the same pool of candidates as the other instances.
+        // 
+        // These not-yet-propagated keys are usually immediately-activated, but an app restarted after a period
+        // of downtime may discover that it has a valid, but soon-to-be-expired key. The replacement will not
+        // be immediately-activated, but may be activated before it has propagated.
+        var useShortRefreshPeriod = generatedKey is not null
+            // Either we had to generate a key and it will be activated before it has time to propagate
+            // (in which case, other instances may have done the same)
+            ? (generatedKey.ActivationDate < now + KeyManagementOptions.KeyPropagationWindow) // No clock skew on a key we generated
+            // Or we selected a key that has yet to propagate (presumably, from another instance)
+            : (defaultKey.CreationDate > now - KeyManagementOptions.KeyPropagationWindow);
```

Comment:
Can this happen in a loop?  What if this is still the default after the next refresh?

Reply:
Indeed it can. 😢